### PR TITLE
IDP-559 - Add Application Insights instrumentation

### PR DESCRIPTION
**Pull request by Gcaya**
## Context
We want to offer teams the possibility to have instrumentation with Azure AppInsights since it is still very much used.

## What was done
- Added new AppInsights project
- Implemented AppInsights behavior for publishing and subscription
- Registered both behaviors in their requried spot within the serviceCollection
- Added a internal wrapper event that is responsible to hold the activity context properties and the original domain event
- Added new tests and fixed existing that were broken following changes

### CODE REVIEW

File: src/DomainEventPropagation/DomainEventWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Text.Json;
+
+namespace Workleap.DomainEventPropagation;
+
+internal class DomainEventWrapper : IDomainEvent
```
Comment (asimmon) [inline src/DomainEventPropagation/DomainEventWrapper.cs:5]: Can we add `JsonPropertyName` attributes with single letter JSON property name in order to reduce the size of the generated JSON? I suggest:

DomainEventType: `t`
DomainEventJson: `j`
ExtensionAttributes: `e`

File: src/Workleap.DomainEventPropagation.ApplicationInsights/TelemetryClientExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Workleap.DomainEventPropagation;
+
+internal static class TelemetryClientExtensions
+{
+    public static IOperationHolder<DependencyTelemetry> StartActivityAwareDependencyOperation(this TelemetryClient telemetryClient, object request)
+    {
+        if (Activity.Current is { } activity && TracingHelper.IsEventGridActivity(activity))
+        {
+            // When the current activity is our own Mediator activity created in our previous request tracing behavior,
```
Comment (asimmon) [inline src/Workleap.DomainEventPropagation.ApplicationInsights/TelemetryClientExtensions.cs:14]:
```suggestion
            // When the current activity is our own Event Grid activity created in our previous event tracing behavior,
```

File: src/Workleap.DomainEventPropagation.ApplicationInsights/SubscriptionApplicationInsightsTracingBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        this._telemetryClient = telemetryClient;
+    }
+
+    public Task Handle(IDomainEvent domainEvent, SubscriberDomainEventsHandlerDelegate next, CancellationToken cancellationToken)
+    {
+        return this._telemetryClient == null ? next(domainEvent) : this.HandleWithTelemetry(domainEvent, next, cancellationToken);
+    }
+
+    private async Task HandleWithTelemetry(IDomainEvent domainEvent, SubscriberDomainEventsHandlerDelegate next, CancellationToken cancellationToken)
```
Comment (asimmon) [inline src/Workleap.DomainEventPropagation.ApplicationInsights/SubscriptionApplicationInsightsTracingBehavior.cs:20]: Cancellation token is not used, so we can remove it from the arguments

File: src/Workleap.DomainEventPropagation.ApplicationInsights/PublishigApplicationInsightsTracingBehavior.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        this._telemetryClient = telemetryClient;
+    }
+
+    public Task Handle(IEnumerable<IDomainEvent> events, DomainEventsHandlerDelegate next, CancellationToken cancellationToken)
+    {
+        return this._telemetryClient == null ? next(events) : this.HandleWithTelemetry(events, next, cancellationToken);
+    }
+
+    private async Task HandleWithTelemetry(IEnumerable<IDomainEvent> events, DomainEventsHandlerDelegate next, CancellationToken cancellationToken)
```
Comment (asimmon) [inline src/Workleap.DomainEventPropagation.ApplicationInsights/PublishigApplicationInsightsTracingBehavior.cs:20]: Cancellation token is not used, so we can remove it from the arguments

File: src/DomainEventPropagation/DomainEventWrapper.cs (hunk 1 above)
Comment (Gcaya) [reply src/DomainEventPropagation/DomainEventWrapper.cs:5]: Yeah good point, will add it!

**Pull request closed without merging**